src/main/java/com/cloudbees/jenkins/GitHubRepositoryNameContributor.java

```diff
@@ -36,41 +37,54 @@
      * then puts them into the collection.
      *
-     * @deprecated Use {@link #parseAssociatedNames(Job, Collection)}
+     * @deprecated Use {@link #parseAssociatedNames(Item, Collection)}
      */
     @Deprecated
     public void parseAssociatedNames(AbstractProject<?, ?> job, Collection<GitHubRepositoryName> result) {
-        parseAssociatedNames((Job) job, result);
+        parseAssociatedNames((Item) job, result);
     }
 
     /**
      * Looks at the definition of {@link Job} and list up the related github repositories,
      * then puts them into the collection.
+     * @deprecated Use {@link #parseAssociatedNames(Item, Collection)}
      */
+    @Deprecated
     public /*abstract*/ void parseAssociatedNames(Job<?, ?> job, Collection<GitHubRepositoryName> result) {
-        if (overriddenMethodHasDeprecatedSignature(job)) {
-            parseAssociatedNames((AbstractProject) job, result);
-        } else {
-            throw new AbstractMethodError("you must override the new overload of parseAssociatedNames");
-        }
+        parseAssociatedNames((Item) job, result);
     }
 
     /**
-     * To select backward compatible method with old extensions
-     * with overridden {@link #parseAssociatedNames(AbstractProject, Collection)}
-     *
-     * @param job - parameter to check for old class
-     *
-     * @return true if overridden deprecated method
+     * Looks at the definition of {@link Item} and list up the related github repositories,
+     * then puts them into the collection.
      */
-    private boolean overriddenMethodHasDeprecatedSignature(Job<?, ?> job) {
-        return Util.isOverridden(
+    @SuppressWarnings("deprecation")
+    public /*abstract*/ void parseAssociatedNames(Item item, Collection<GitHubRepositoryName> result) {
+        if (Util.isOverridden(
+                GitHubRepositoryNameContributor.class,
+                getClass(),
+                "parseAssociatedNames",
+                Job.class,
+                Collection.class
+        )) {
+            // if this impl is legacy, it cannot contribute to non-jobs, so not an error
+            if (item instanceof Job) {
+                parseAssociatedNames((Job<?, ?>) item, result);
+            }
+        } else  if (Util.isOverridden(
                 GitHubRepositoryNameContributor.class,
                 getClass(),
                 "parseAssociatedNames",
                 AbstractProject.class,
                 Collection.class
-        ) && job instanceof AbstractProject;
+        )) {
+            // if this impl is legacy, it cannot contribute to non-projects, so not an error
+            if (item instanceof AbstractProject) {
+                parseAssociatedNames((AbstractProject<?, ?>) item, result);
+            }
+        } else {
+            throw new AbstractMethodError("you must override the new overload of parseAssociatedNames");
```

Comment:
Actually this is more a design error in the original API. These methods taking the collection should actually have been `protected` as they form part of the SPI not the API.

If you have an old implementation, then you will have overridden `parseAssociatedNames(Job,Collection)` or `parseAssociatedNames(AbstractProject,Collection)` and thus we delegate to those.

If you have a new implementation, then you are supposed to override `parseAssociatedNames(Item,Collection)` and if you don't then this exception should blow up in your face and give you the AME.

If we had these SPI methods as `protected` then we could make `parseAssociatedNames(Item,Collection)` abstract and use the API methods like `parseAssociatedNames(Item)` to route through to the implementation of existing extension points... but instead, as this is public API we have to do it this way and we have no way to force new implementations to override `parseAssociatedNames(Item,Collection)` other than this exception